File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
      if len(s) % 2 > 0: return False
      stack = []
      for ch in s:
        if ch == '(': stack += ')'
        elif ch == '[': stack += ']'
        elif ch == '{': stack += '}'
        else:
          if len(stack) == 0 or stack.pop() != ch:
            return False
        
      return len(stack) == 0
    
        

File: 20-valid-parentheses/20-valid-parentheses.cs
public class Solution {
  private static Dictionary<char, char> map = new () {
    {'{','}'}, {'[',']'}, {'(',')'}
  };
  public bool IsValid(string s) {
    Stack<char> stack = new();
    foreach (char ch in s) {
      if (map.ContainsKey(ch)) {
        stack.Push(map[ch]);
      }
      else {
        if (stack.Count == 0 || stack.Pop() != ch)
          return false;
      }
    }
    return stack.Count == 0;
  }
}